### ISSUE TITLE
[question] Should we allow timeouts and other similar settings to be configured by users?

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#986]
From https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/327#note_204942

@dbeltrankyl I thought about this issue on my way home yesterday, and realized that maybe this should be further considered. Especially now that we will have Autosubmit running on LUMI (and possibly other platforms like Juwels, Levante, Meluxina, etc.).

From what I understood, if one of these sites need to increase the timeout for issues such as network glitch, slowness on the other end, operating system bug, etc., then they would only have two options: a) hack the local source code of the installed Python package for Autosubmit, or b) file an issue to us and wait for a new release where we have changed the setting that's hard-coded our the source code.

I believe that would be frustrating for these users, and they would probably expect to not have to ask us for changing the timeouts in their environment. Furthermore, if we changed that to a very high value, that would impact all the installations using this new version, not only the one from the user that requested it.

I think it could be a setting somewhere in our YAML configuration, maybe under the `CONFIG` key. That way users would be free to use the default value (which we can set to `60` or something we consider a sensible default), and instruct users on our docs how to configure it.

WDYT?

Here's the docs for timeouts (in different contexts) used in other WMS:

- ecFlow (on the client side via `ECF_TIMEOUT` env var): https://ecflow.readthedocs.io/en/latest/python_api/Client.html
- ecFlow (via Troika, setting SSH timeout): https://github.com/ecmwf/troika/blob/faf6fd64148fa5f16ac756bce521d181008f1dc6/docs/reference/config.rst#ssh_connect_timeout
- Cylc (they have timeouts in other configuration sections too, like inactivity timeout, submission timeout, etc.): https://cylc.github.io/cylc-doc/stable/html/user-guide/task-implementation/job-submission.html#execution-time-limit-and-execution-timeout
- Pegasus (the WMS that was presented at the BSC this week :nerd:, see their entry 6. in the changelog - their docs are a bit confusing IMO): https://pegasus.isi.edu/2015/08/10/pegasus-4-5-1released/
- Nextflow (only found in their AWS scope, to control connection and socket timeout): https://www.nextflow.io/docs/latest/config.html#scope-aws
- COMPSs (see their `@ task(time_out=NN)` parameter): https://compss-doc.readthedocs.io/en/stable/Sections/02_App_Development/02_Python/01_1_Task_definition/Sections/04_Task_parameters_summary.html

**Comment by kinow** [comment]
To clarify, this is not only about the ssh connection timeout, but about other timeout and similar settings that we have hard-coded in our code base. Anything that could be different in, for instance, MN4 and LUMI.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 22, 2023, 11:29

Hello @kinow , 

The current timers are: 

- Connect: banner, socket 60seconds
- Transport: 60 seconds ( new one, yet to see if it will work)
- Ftp channel ( get files, send files ) : Couldn't set a timeout for FTP itself
- Commands: Rm = 60 seconds, submit, etc. 120 seconds. Migrate = infinite, X11 = infinite.
- By default, Autosubmit more than once each command submitted if it has issues. And has an incremental delay ( I think up to 3 times ) 
- Recovery: 15-120seconds between retrials, max retrials 72h-122h ( adjustable this last part)  
- timeout between autosubmit iterations 10 seconds. 
- timeout between wrapper checks 10 seconds.


Then each command has a timeout between retrials of 5-30 seconds and three retrials ( If I recall correctly ) 

Ajustable parameters:

timeout between autosubmit iterations, 
timeout between wrapper checks

I don't see any issue adding adjustable parameters if they don't mess with migrate command or x11. 

My only concern with allowing to increase in the number of retrials or timeouts is due to the current log_recovery threads. I would address that performance issue before we allow to increment the timeout because we could have memory issues if the amount of threads recovery jobs are large enough. 

In summary, if we solve #554 and don't touch timeout for migrate/ x11 I don't see an issue with it

**Comment by kinow** [comment]
Thanks for the detailed reply @dbeltrankyl !

>I don't see any issue adding adjustable parameters if they don't mess with migrate command or x11.

+1

>My only concern with allowing to increase in the number of retrials or timeouts is due to the current log_recovery threads. I would address that performance issue before we allow to increment of the timeout because we could have memory issues if the amount of threads recovery jobs are large enough.
>
>In summary, if we solve #554 and don't touch timeout for migrate/ x11 I don't see an issue with it

:thumbsup: thanks for linking that issue! Will take a look at it later to try to understand what was reported there.

Thanks again!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 22, 2023, 11:39

> thanks for linking that issue! Will take a look at it later to try to understand what was reported there.

Shortly right now, the amount of threads recovering jobs is 1thread:1job. ( and each thread is connecting to the platform etc )  

The idea is to reduce that amount to X based on totaljobs or some parameter of the experiment configuration to reduce the quantity. These master threads would be the ones in charge of recovering the logs* of more than one job. to reduce the overhead drastically. 

The problem is that this implies a refactor in some classes, and I didn't have time . This thread function was done in a moment where the entire autosubmit was stopped for each log to recover ( so we couldn't have more than a hundred jobs )

**Comment by kinow** [comment]
mentioned in issue digital-twins/de_340/workflow#301
